lib/Dialect/ATen/ATenLoweringPass.cpp

```diff
@@ -0,0 +1,972 @@
+  } else if (FloatType ft = ty.dyn_cast<FloatType>()) {
+    // ret << "F" << ft.getWidth();
+  } else if (const IntegerType it = ty.dyn_cast<const IntegerType>()) {
+    // ret << "I" << it.getWidth();
+  } else if (const mlir::NPCOMP::aten::ATenListType alt =
+                 ty.dyn_cast<const mlir::NPCOMP::aten::ATenListType>()) {
+
+  } else {
+    Type t = ty;
+    t.dump();
+    assert(0 && "unhandled type in getSimplyMangledType");
+  }
+  return ret.str();
+}
+
+// Return a mangled function name.  The function name is constructed
+// from the prefix, the mangled result types, the mangled operand types.
+static std::string getSimplyMangledFuncName(std::string prefix,
+                                            ArrayRef<Type> operTy,
+                                            ArrayRef<Type> resultTy) {
+  std::string sep = "_";
+
+  std::string ret = prefix;
+  for (const Type t : resultTy)
+    ret = ret + sep + getSimplyMangledType(t);
+  for (const Type t : operTy) {
+    std::string s = getSimplyMangledType(t);
+    if(s.size() > 0)
+      ret = ret + sep + getSimplyMangledType(t);
+  }
+  ret += "_out";
+
+  return ret;
+}
+static std::string getSimplyMangledFuncName(std::string prefix,
+                               FunctionType fnTy) {
+
+  auto resultTy = fnTy.getResults();
+  auto operTy = fnTy.getInputs();
+  return getSimplyMangledFuncName(prefix, operTy, resultTy);
+}
+
+std::string getMangledFuncName(std::string prefix,
+                               FunctionType fnTy) {
+  //return getFullyMangledFuncName(prefix, fnTy);
+  return getSimplyMangledFuncName(prefix, fnTy);
+}
+
+std::string getMangledFuncName(std::string prefix,
+                               ArrayRef<Type> opTys,
+                               ArrayRef<Type> retTys) {
+  //return getFullyMangledFuncName(prefix, fnTy);
+  return getSimplyMangledFuncName(prefix, opTys, retTys);
+}
+
+static FuncOp getATenFn(ModuleOp module, std::string mangledFunctionName,
+                        ArrayRef<Value> operands,
+                        ArrayRef<Type> retTys) {
+  Builder builder(module);
+
+  SmallVector<Type, 8> tys;
+  for (auto o : operands) {
+    Type t = o.getType();
+    // Erase the dimensions of the memref.
+    if (t.isa<MemRefType>()) {
+      auto mt = t.cast<MemRefType>();
+      tys.push_back(getShapeErasedMemRefType(mt));
+    } else
+      tys.push_back(t);
+  }
+
+  auto fnTy = builder.getFunctionType(tys, retTys);
+
+  auto fn = module.lookupSymbol<FuncOp>(mangledFunctionName);
+
+  if (!fn) {
+    fn = FuncOp::create(builder.getUnknownLoc(), mangledFunctionName, fnTy);
+    module.push_back(fn);
+  }
+
+  return fn;
+}
+
+/// Lower an aten.add to an affine loop nest.
+///
+/// This class inherit from `ConversionPattern` and override `rewrite`,
+/// similarly to the PatternRewriter introduced in the previous chapter.
+/// It will be called by the DialectConversion framework (see `ATenLowering`
+/// class below).
```

Comment:
This looks like a partial lowering for add. Can you document what isn't covered?